[DA] handle memory accesses with different offsets and strides

sebpop opened the pull request:
This patch corrects the behavior of the Dependence Analysis for memory accesses that do not start at the same offset or do not have similar strides. When offsets or strides cannot be disambiguated at compile time, DA collects a set of runtime assumptions under which the dependence test becomes valid. The default remains the same as before the patch: DA rejects the dependence test as undecidable instead of collecting runtime assumptions.

sebpop committed:
[DA] Dependence analysis does not handle array accesses of different sizes

This fixes bug https://github.com/llvm/llvm-project/issues/16183
where the elements of a same array are accesses as i32 and i64.
This is not handled by the array dependence analysis.

sebpop committed:
[DA] check memory offsets are multiples of elements size

Check that the offsets from memory base address are multiples of the array
element size for a pair of memory accesses in a dependence test.

The patch adds a testcase that cannot be disambiguated by the DA.

Meinersbur commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3625,21 +3621,22 @@ DependenceInfo::depends(Instruction *Src, Instruction *Dst) {
-
-  FullDependence Result(Src, Dst, PossiblyLoopIndependent, CommonLevels);
-  ++TotalArrayPairs;
+  if (DstLoc.Size != SrcLoc.Size) {
```
This is #116630?

sebpop replied:
Correct.  This patch is on top of #116630.
I can either rebase this PR on top of nothing, or squash both patches in one.
I would prefer to separately commit #116630 first, and then work on to this PR.

Meinersbur commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3649,6 +3646,111 @@ DependenceInfo::depends(Instruction *Src, Instruction *Dst) {
     return std::make_unique<Dependence>(Src, Dst);
   }
+
+  auto EltSize = SrcLoc.Size.toRaw();
```
[style] [No Almost-Always-Auto](https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable)

Meinersbur commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3649,6 +3646,111 @@ DependenceInfo::depends(Instruction *Src, Instruction *Dst) {
+  std::function<bool(const SCEV *, const SCEV *&)> checkOffsets =
+      [&](const SCEV *V, const SCEV *&Param) -> bool {
+    if (auto *AddRec = dyn_cast<SCEVAddRecExpr>(V)) {
+      if (!checkOffsets(AddRec->getStart(), Param))
```
For recusive calls, I'd avoid lambdas. The necessary cast to `std::function` causes a lot of overhead that is unlikely to be optimized aways.

Meinersbur commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3649,6 +3646,111 @@ DependenceInfo::depends(Instruction *Src, Instruction *Dst) {
+      return C.srem(EltSize) == 0;
+    }
+
+    auto checkParamsMultipleOfSize = [&](const SCEV *V,
```
Lambda-within-lamda?

Meinersbur commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3649,6 +3646,111 @@ DependenceInfo::depends(Instruction *Src, Instruction *Dst) {
+    }
+  }
+
+  // establish loop nesting levels
```
Useless comment

Meinersbur commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3649,6 +3646,111 @@ DependenceInfo::depends(Instruction *Src, Instruction *Dst) {
+  auto EltSize = SrcLoc.Size.toRaw();
+  assert(EltSize == DstLoc.Size.toRaw() && "Array element size differ");
+
+  // Check that memory access offsets in V are multiples of array EltSize.
```
Please elaborate more. What is the assumed base ptr? What is is `Param`. Where does EltSize come from?

sebpop committed:
fix auto type

sebpop committed:
convert checkOffsets lambda to static function

sebpop committed:
remove useless comment

sebpop committed:
add more comments and examples

Meinersbur commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3569,6 +3569,123 @@ bool DependenceInfo::invalidate(Function &F, const PreservedAnalyses &PA,
          Inv.invalidate<LoopAnalysis>(F, PA);
 }
 
+// Check that memory access offsets in V are multiples of array element size
+// EltSize. Param records the first parametric expression. If the scalar
+// evolution V contains two or more parameters, we check that the subsequent
+// parametric expressions are multiples of the first parametric expression
+// Param.
```
```suggestion
/// Check that memory access offsets in V are multiples of array element size
/// EltSize. Param records the first parametric expression. If the scalar
/// evolution V contains two or more parameters, we check that the subsequent
/// parametric expressions are multiples of the first parametric expression
/// Param.
```

Meinersbur commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3569,6 +3569,123 @@ bool DependenceInfo::invalidate(Function &F, const PreservedAnalyses &PA,
+static bool checkOffsets(ScalarEvolution *SE, const SCEV *V, const SCEV *&Param,
+                         uint64_t EltSize) {
+  if (auto *AddRec = dyn_cast<SCEVAddRecExpr>(V)) {
+    if (!checkOffsets(SE, AddRec->getStart(), Param, EltSize))
+      return false;
+    return checkOffsets(SE, AddRec->getStepRecurrence(*SE), Param, EltSize);
```
```suggestion
    return checkOffsets(SE, AddRec->getStart(), Param, EltSize) &&
      checkOffsets(SE, AddRec->getStepRecurrence(*SE), Param, EltSize);
```
If you think the current is more readable, that's fine too.

Meinersbur commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3569,6 +3569,123 @@ bool DependenceInfo::invalidate(Function &F, const PreservedAnalyses &PA,
+// evolution V contains two or more parameters, we check that the subsequent
+// parametric expressions are multiples of the first parametric expression
+// Param.
+static bool checkOffsets(ScalarEvolution *SE, const SCEV *V, const SCEV *&Param,
```
Name suggestion: `isKnownMultipleOf`, analogous to e.g. `ScalarEvoluton::isKnownToBeAPowerOfTwo`. Maybe even move it to be a member of ScalarEvolution?

Meinersbur commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3569,6 +3569,123 @@ bool DependenceInfo::invalidate(Function &F, const PreservedAnalyses &PA,
+
+    // Check whether "(Param - V) % Size == 0".
+    const SCEV *Diff = SE->getMinusSCEV(Param, V);
+    if (const SCEVConstant *Cst = dyn_cast<SCEVConstant>(Diff)) {
```
```suggestion
    if (auto *Cst = dyn_cast<SCEVConstant>(Diff)) {
```

Meinersbur commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3569,6 +3569,123 @@ bool DependenceInfo::invalidate(Function &F, const PreservedAnalyses &PA,
+      APInt Val = Cst->getAPInt();
+      if (Val.isZero())
+        return true;
+      auto Rem = Val.srem(APInt(Val.getBitWidth(), EltSize, /*isSigned=*/true));
```
```suggestion
      APInt Rem = Val.srem(APInt(Val.getBitWidth(), EltSize, /*isSigned=*/true));
```
Shouldn't it be `urem`?

There are overloads of `srem`/`urem` that accept int64_t

Meinersbur commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3569,6 +3569,123 @@ bool DependenceInfo::invalidate(Function &F, const PreservedAnalyses &PA,
+    return checkParamsMultipleOfSize(V, Param);
+
+  // Expressions like "n + 1".
+  if (isa<SCEVAddExpr>(V))
+    return !SCEVExprContains(V, [](const SCEV *S) {
+      return isa<SCEVUnknown>(S);
+    }) || checkParamsMultipleOfSize(V, Param);
+
+  // Expressions like "n * 2".
+  if (isa<SCEVMulExpr>(V))
+    return !SCEVExprContains(V, [](const SCEV *S) {
+      return isa<SCEVUnknown>(S);
+    }) || checkParamsMultipleOfSize(V, Param);
```
```suggestion
  if (isa<SCEVAddExpr>(V) || isa<SCEVMulExpr>(V))
    return !SCEVExprContains(V, [](const SCEV *S) {
      return isa<SCEVUnknown>(S);
    }) || checkParamsMultipleOfSize(V, Param);
```

Why must it not contain a SCEVUnknown?

Meinersbur commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3569,6 +3569,123 @@ bool DependenceInfo::invalidate(Function &F, const PreservedAnalyses &PA,
+      return true;
+    if (!Param) {
+      Param = V;
+      return true;
```
I don't understand why if there is only a single SCEVUnknown involved, why is it automatically a multiple of `EltSize`? 

```llvm
  %ptr = getelementptr i8, ptr %base, i32 %offset
  store i64 42, ptr %ptr 
```
So the array is elements of 8 bytes, but the gep is incrementing by single bytes. If the SCEVUnknown %offset is not a multiple of 8, the check should fail?

sebpop replied:
If a parameter (ssa name invariant in loop) name %n is used as offset, then we assume that %n is a multiple of the array element size.  If another parameter %m is also used as an offset of the same array, then the test fails because we do not know the relation between %n and %m.
We could collect all those parametric expressions and implement a runtime check (%n % 8 == 0) && (%m % 8 == 0) with assumptions like we do in Polly, although the dependence analysis will only provide the set of constraints as a context, and it should be the loop transforms to version the code according to validity constraints.

Meinersbur replied:
> If a parameter (ssa name invariant in loop) name %n is used as offset, then we assume that %n is a multiple of the array element size. There isn't even a check that `%n` is loop-invariant.

That ssumption that obviously does not necessarily hold true.

I think what you mean is that DA is assuming that `%n` is part of all index expressions and therefore has the same offset everywhere. If some index expressions use `%m` instead, they may have different relative offsets. 

I do not think this is sound. Some index expressions may also use `%n + %n`, `%n + 1`, or no `%n` at all in which case they may have different relative offsets already. Alternatively, I may not have understood the logic here.

If `%n` (or any other subexpression) is invariant to all loops and occurs in every index, I think it could considered part of the base pointer. `ScalarEvolution::getPointerBase` atm looks into all SCEVAddExpr disregarding whether operands are loop invariant. Maybe that's a viable approach.

sebpop replied:
> DA is assuming that %n is part of all index expressions and therefore has the same offset everywhere. If some index expressions use %m instead, they may have different relative offsets.

You are right.  If both %n and %m are used to index elements in a same array, the difference `%n - %m` is unknown at compilation time, and so the current dependence analysis result would be incorrect.

> Some index expressions may also use %n + %n, %n + 1, or no %n at all in which case they may have different relative offsets already.

Correct.
This patch tries to prove at compile time that if an array offset is using %n, then all other array offsets will use %n in multiples of the array element size.

For example, if eltSize = 8, and (%n mod 8) = x then ((%n + 1) mod 8) = x + 1. 
Even if we do not know the value of x at compilation time, we know that 2 arrays accessing %n and %n + 1 do not overlap correctly (not a multiple of 8) for the dependence test to produce a valid answer.

> If %n (or any other subexpression) is invariant to all loops and occurs in every index, I think it could be considered part of the base pointer.

Correct.

sjoerdmeijer replied:
> > DA is assuming that %n is part of all index expressions and therefore has the same offset everywhere. If some index expressions use %m instead, they may have different relative offsets.
> 
> You are right. If both %n and %m are used to index elements in a same array, the difference `%n - %m` is unknown at compilation time, and so the current dependence analysis result would be incorrect.

I am catching up on this, trying to let this sink in, but had this perhaps silly question in the meantime: if there are still correctness issues, can we not just return Confused for now when we detect that we have different SCEV expressions with different offsets? Or would this make DA so conservative that it will be useless?

sebpop replied:
Out of 43 in test/Analysis/DependenceAnalysis half are failing if I disable DA when the array access uses a parameter: base + parameter.

sebpop committed:
add extra slash for function comments

Co-authored-by: Michael Kruse <github@meinersbur.de>

sebpop committed:
simplify logic with a single return

Co-authored-by: Michael Kruse <github@meinersbur.de>

sebpop committed:
use auto on dyn_cast assign for readability

Co-authored-by: Michael Kruse <github@meinersbur.de>

sebpop committed:
simplify logic for similar code-paths

Co-authored-by: Michael Kruse <github@meinersbur.de>

sebpop committed:
move isKnownMultipleOf to ScalarEvolution

kasuga-fj commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -10983,6 +10983,110 @@ bool ScalarEvolution::isKnownToBeAPowerOfTwo(const SCEV *S, bool OrZero,
+  // expressions are multiples of the first parametric expression Param.
+  auto checkParamsMultipleOfSize = [&](const SCEV *V,
+                                       const SCEV *&Param) -> bool {
+    if (EltSize == 1)
+      return true;
```
Can move to outside of the lambda?

kasuga-fj commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -10983,6 +10983,110 @@ bool ScalarEvolution::isKnownToBeAPowerOfTwo(const SCEV *S, bool OrZero,
+    if (Param == V)
+      return true;
+
+    // Check whether "(Param - V) % Size == 0".
```
```suggestion
    // Check whether "(Param - V) % EltSize == 0".
```

kasuga-fj commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -10983,6 +10983,110 @@ bool ScalarEvolution::isKnownToBeAPowerOfTwo(const SCEV *S, bool OrZero,
+                        << Rem << " != 0\n");
+      return false;
+    }
+    // Check if the symbolic difference is a multiple of Size.
```
```suggestion
    // Check if the symbolic difference is a multiple of EltSize.
```

kasuga-fj reviewed (commented):
I'm not familiar with DependenceAnalysis, but I've read the code and will comment on it. From what I've understood, I think what @Meinersbur said is correct.

This is just my guess, but does this patch tries to guarantee correctness based on the fact that `SrcPtr` and `DstPtr` are congruent modulo `EltSize` if all index expressions are multiples of the same value (`%n`)? I haven't thought about this enough to be certain that this approach is correct, but even if it is, I don't think it is properly implemented.

I have one more question. Is it insufficient to check the `align` value is equal to `EltSize` for both `Src` and `Dst`?

sebpop committed:
record runtime assumptions for parametric expressions

sebpop committed:
disable loop interchange, fusion, and unroll-and-jam on runtime assumptions

kasuga-fj reviewed (commented):
I'll check the discourse later.

sebpop commented on llvm/lib/Transforms/Scalar/LoopInterchange.cpp:
```diff
@@ -1834,6 +1838,12 @@ PreservedAnalyses LoopInterchangePass::run(LoopNest &LN,
   std::unique_ptr<CacheCost> CC =
       CacheCost::getCacheCost(LN.getOutermostLoop(), AR, DI);
 
+  SCEVUnionPredicate Assumptions = DI.getRuntimeAssumptions();
+  // Early fail when the dependence analysis has runtime assumptions.
+  // FIXME: this could be handled by versioning the loop.
+  if (!Assumptions.isAlwaysTrue())
+    return PreservedAnalyses::all();
```
Just above this is a call to CacheCost::getCacheCost that queries all dependences in the loop nest to compute cache reuse patterns.

kasuga-fj replied:
It's true, I missed it.

kasuga-fj replied:
Hmm, so does this mean that the user of `CacheCostAnalysis` must check if DA assumes some conditions? I think it's inconvenient, I'd like an analysis pass to do as much as it can without any additional assumptions.

kasuga-fj replied:
> I think it's inconvenient, I'd like an analysis pass to do as much as it can without any additional assumptions by default.

Can you consider this? For example, I think there is a way to provide some API to control the default policy that `DependenceInfo` follows to determine if it is allowed to append runtime checks.

sebpop commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -10971,6 +10971,67 @@ bool ScalarEvolution::isKnownToBeAPowerOfTwo(const SCEV *S, bool OrZero,
+
+  // Expressions like "n + 1" and "n * 3".
+  if (isa<SCEVAddExpr>(S) || isa<SCEVMulExpr>(S)) {
+    if (SCEVExprContains(S, [](const SCEV *X) { return isa<SCEVUnknown>(X); }))
+      recordRuntimePredicate(S);
+    return true;
```
You're right, I will amend this part.

sebpop commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -10971,6 +10971,67 @@ bool ScalarEvolution::isKnownToBeAPowerOfTwo(const SCEV *S, bool OrZero,
+        getURemExpr(S, getConstant(ConstantInt::get(STy, M, false)));
+    const SCEV *Zero = getZero(STy);
+
+    // Check whether "S % M == 0" is known at compile time.
+    if (isKnownPredicate(ICmpInst::ICMP_EQ, SmodM, Zero))
+      return;
```
You're right, I will amend this part.

kasuga-fj commented:
I don't fully understand this component yet, but not sure if "the offset is `%N` and `%N % 4 == 0"` and "the offset is `4 * %n` where `%N == 4 * %n` are equivalent. That is, if `isKnownMultipleOf` requires some assumptions, would it be better to replace the offset `%N` with `4 * %n` where `%N == 4 * %n`?

sebpop committed:
handle compile time "s % m != 0" in isKnownMultipleOf

Do not check for symbolic names, just handle all possible SCEV expressions with
compile time tests and record runtime predicates when compile time fails.

sebpop committed:
record runtime predicates on each Dependence relation

sebpop committed:
add testcase from Ryotaro Kasuga's review

sebpop commented:
> I don't fully understand this component yet, but not sure if "the offset is `%N` and `%N % 4 == 0"` and "the offset is `4 * %n` where `%N == 4 * %n` are equivalent. For example, if we have the following IR
> 
> ```
> %arrayidx0 = getelementptr inbounds i8, ptr %A, i64 %N
> store i32 42, ptr %arrayidx0, align 1
> 
> %arrayidx1 = getelementptr inbounds i32, ptr %A, i64 %M
> store i32 42, ptr %arrayidx1, align 4
> ```

I added this as a testcase in 480ceff.

> 
> Under the assumption `%N % 4 == 0`, the former means like `A[N/4] = 42`, and the later is `A[M] = 42`. It seems that the current implementation doesn't perform such explicit division. 

DA sees the following access functions:
```
    subscript 0
	src = (%n + %A)
	dst = ((4 * %m)<nsw> + %A)
```
The compile time check for `4 * %m modulo 4 == 0` is trivial, only a runtime check for `%n modulo 4 == 0` is recorded.

kasuga-fj commented:
> > I don't fully understand this component yet, but not sure if "the offset is %N and %N % 4 == 0" and "the offset is 4 * %n where %N == 4 * %n are equivalent. For example, if we have the following IR
> > ```
> > %arrayidx0 = getelementptr inbounds i8, ptr %A, i64 %N
> > store i32 42, ptr %arrayidx0, align 1
> >
> > %arrayidx1 = getelementptr inbounds i32, ptr %A, i64 %M
> > store i32 42, ptr %arrayidx1, align 4
> > ```
> I added this as a testcase in https://github.com/llvm/llvm-project/commit/480ceff66fdf8491aeb695ddd08107d92b3f8d23.

I may have confused you, but I don't think this example directly causes any problems. What I'm concerned about is a case where a target loop has such a code fragment in its body. Looking at the testcases of Delinearization, the subscript for the second store can become `%m`, not `4 * %m`?

https://github.com/llvm/llvm-project/blob/0f526498a1b0819055d1094fd9850fdbcab5b903/llvm/test/Analysis/Delinearization/multidim_ivs_and_parameteric_offsets_3d.ll

If so, what happens if Delinearization succeeded only for the second store? I don't have any cases to check for such a situation, but I guess the subscripts would become like the following.

```
subscript 0
	src = (%n + %A)
	dst = (%m + %A)
```

In this case the analysis still works fine, because we don't know the relation between `%n` and `%m`, so they are handled conservatively. But what about the case where `%m` is a function of `%n`, like `%n + 1`?

kasuga-fj commented on llvm/include/llvm/Analysis/DependenceAnalysis.h:
```diff
@@ -76,8 +76,12 @@ namespace llvm {
-    Dependence(Instruction *Source, Instruction *Destination)
-        : Src(Source), Dst(Destination) {}
+    Dependence(Instruction *Source, Instruction *Destination,
+               ScalarEvolution *SE,
```
nit: Passing `SE` when calling `getRuntimeAssumptions` may be better.

sebpop commented:
> But what about the case where %m is a function of %n, like %n + 1?

If you have a testcase that shows the issue, I will add it to the tests.
I believe I addressed all your previous comments. Thanks for your reviews.

kasuga-fj commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3617,30 +3633,80 @@ DependenceInfo::depends(Instruction *Src, Instruction *Dst) {
+    if (!UnderRuntimeAssumptions)
+      return std::make_unique<Dependence>(Src, Dst,
+                                          SCEVUnionPredicate(Assume, *SE));
+    if (Assumptions.empty()) {
+      Assumptions.append(Assume.begin(), Assume.end());
```
Perhaps this branch is unnecessary.

sebpop replied:
I don't follow.  If the Assumptions is empty, then we just copy everything from Assume.  However if Assumptions is non-empty, we only add the non-redundant assumptions.

kasuga-fj replied:
I still think this branch is unnecessary.

kasuga-fj commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3617,30 +3633,80 @@ DependenceInfo::depends(Instruction *Src, Instruction *Dst) {
+    if (Assumptions.empty()) {
+      Assumptions.append(Assume.begin(), Assume.end());
+    } else {
+      // Add non-redundant assumptions.
```
(This is just a comment, not a request to fix anything.)
I realized that `Assume` might contain redundant predicts, e.g.,
```
EltSize = 4
SrcSCEV = 2 * %n
DstSCEV = %n
```

First we will call `isKnownMultipleOf(SrcEV, EltSize, Assume)` and it will add a predicate for `(2 * n) % 4 == 0`. Next `isKnownMultipleOf(DstEV, EltSize, Assume)` will be called and it will add `n % 4 == 0`, which is stricter than the previous one.
In principle, if `SCEVPredicate::implies` is complete, the current algorithm allows to add a predicate to Assume that might be stricter than the one previously added, but if I understand correctly

kasuga-fj reviewed (commented):
> If you have a testcase that shows the issue, I will add it to the tests.

OK, I will think about it some more.

kasuga-fj commented:
(NOTE: I think you should avoid force-pushing unless rebasing is necessary. Please read [LLVM GitHub User Guid](https://llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes) for more details.)

kasuga-fj commented:
I was concerned about like the following case.

```
define void @foo(i64 %n, i64 %m, i64 %o, ptr %A) {
entry:
  br label %for.i

for.i:
  %i = phi i64 [ 0, %entry ], [ %i.inc, %for.i.inc ]
  br label %for.j

for.j:
  %j = phi i64 [ 0, %for.i ], [ %j.inc, %for.j.inc ]
  br label %for.k

for.k:
  %k = phi i64 [ 0, %for.j ], [ %k.inc, %for.k.inc ]
  %subscript0 = mul i64 %i, %m
  %subscript1 = add i64 %j, %subscript0
  %subscript2 = mul i64 %subscript1, %o
  %subscript3 = add i64 %subscript2, %k
  %idx0 = getelementptr inbounds i64, ptr %A, i64 %subscript3  ; (i64*)(A) + i*m*o + j*o + k
  store i32 1, ptr %idx0
  %idx1 = getelementptr inbounds i32, ptr %A, i64 %subscript3  ; (i32*)(A) + i*m*o + j*o + k
  store i32 1, ptr %idx1
  br label %for.k.inc

for.k.inc:
  %k.inc = add nsw i64 %k, 1
  %k.exitcond = icmp eq i64 %k.inc, %o
  br i1 %k.exitcond, label %for.j.inc, label %for.k

for.j.inc:
  %j.inc = add nsw i64 %j, 1
  %j.exitcond = icmp eq i64 %j.inc, %m
  br i1 %j.exitcond, label %for.i.inc, label %for.j

for.i.inc:
  %i.inc = add nsw i64 %i, 1
  %i.exitcond = icmp eq i64 %i.inc, %n
  br i1 %i.exitcond, label %end, label %for.i

end:
  ret void
}
```

I tried this case, then the delinearization itself succeeded and the subscripts for both stores became as follows

```
Subscripts:
{0,+,1}<nuw><nsw><%for.i>
{0,+,1}<nuw><nsw><%for.j>
{0,+,1}<nuw><nsw><%for.k>
```

But it was rejected by delinearization check in DA, and the result of dependency analysis was correct.

https://github.com/llvm/llvm-project/blob/3dbba8e1ab86cc1b1b7514ad3576077f9f24dca8/llvm/lib/Analysis/DependenceAnalysis.cpp#L3452-L3469

A slight modification of the above code, as shown below, bypasses this check and returns incorrect results. I'm not confident if this IR is legal, but if it is, I think we need to make some kind of fix for the delinearization. So, my question is that, is the below LLVM IR valid?

Input:
```
define void @foo(ptr %A) {
entry:
  br label %for.i

for.i:
  %i = phi i64 [ 0, %entry ], [ %i.inc, %for.i.inc ]
  br label %for.j

for.j:
  %j = phi i64 [ 0, %for.i ], [ %j.inc, %for.j.inc ]
  br label %for.k

for.k:
  %k = phi i64 [ 0, %for.j ], [ %k.inc, %for.k.inc ]
  %idx0 = getelementptr inbounds [256 x [256 x [256 x i64]]], ptr %A, i64 %i, i64 %j, i64 %k
  store i32 1, ptr %idx0
  %idx1 = getelementptr inbounds [256 x [256 x [256 x i32]]], ptr %A, i64 %i, i64 %j, i64 %k
  store i32 1, ptr %idx1
  br label %for.k.inc

for.k.inc:
  %k.inc = add nsw i64 %k, 1
  %k.exitcond = icmp eq i64 %k.inc, 256
  br i1 %k.exitcond, label %for.j.inc, label %for.k

for.j.inc:
  %j.inc = add nsw i64 %j, 1
  %j.exitcond = icmp eq i64 %j.inc, 256
  br i1 %j.exitcond, label %for.i.inc, label %for.j

for.i.inc:
  %i.inc = add nsw i64 %i, 1
  %i.exitcond = icmp eq i64 %i.inc, 256
  br i1 %i.exitcond, label %end, label %for.i

end:
  ret void
}
```

Output:
```
Src:  store i32 1, ptr %idx0, align 4 --> Dst:  store i32 1, ptr %idx1, align 4
  da analyze - consistent output [0 0 0|<]!
```

sjoerdmeijer commented:
This patch has grown into doing at least 3 things, I think:
- it adds the concepts of "assumptions" to DA,
- it fixes offset handling in DA,
- and it fixes up interchange, and unroll-and jam.

I guess it would be easier to untangle this and create separate (stacked) patches for this, so that it's easier to review and maintain (in case we would need to partially revert/recommit some of this later). E.g., it looks like the assumptions addition can be tested in isolation as I see support for it being added to dump functions, i.e. we don't need the offset fix for that. 
Does this make sense?

kasuga-fj commented:
> I guess it would be easier to untangle this and create separate (stacked) patches for this, so that it's easier to review and maintain (in case we would need to partially revert/recommit some of this later). E.g., it looks like the assumptions addition can be tested in isolation as I see support for it being added to dump functions, i.e. we don't need the offset fix for that.
Does this make sense?

I agree with this. I think this PR started out trying to handle different stride/load/store sizes as much as possible without runtime assumptions, but the approach and implementation has changed and now contains several features that should be split into several patches. IMHO, I prefer the previous approach, i.e., rejecting cases where we need one or more runtime checks. The current implementation looks better than the previous one, so as the first step, it seems reasonable to me to give up cases where we need one or more runtime assumptions (i.e., `UnderRuntimeAssumptions` should be `false` by default). If we do this, we don't need to fix each client of DA.

Note that, if you want `UnderRuntimeAssumptions` to be `true` by default, the fixes for the users of DA (interchange, fuse, etc.) cannot be separated into different patches, this would break correctness.

sebpop committed:
add testcases

sebpop committed:
Revert "disable loop interchange, fusion, and unroll-and-jam on runtime assumptions"

This reverts commit 8c69ebf95b860081c084b542f107ccbc000afc8f.

sebpop committed:
turn off UnderRuntimeAssumptions in depends computation

sebpop committed:
clang-format code around my changes

sjoerdmeijer reviewed (approved):
I think this is an elegant way of moving this forward. The assumptions are off by default, so won't disturb other users. Other folks have looked at other parts of this patch, I am happy with the assumptions, so this LGTM. But please leave enough time for others to comment though, at least a couple of days, before merging this.

kasuga-fj commented on llvm/include/llvm/Analysis/ScalarEvolution.h:
```diff
@@ -1044,6 +1044,11 @@ class ScalarEvolution {
   bool isKnownToBeAPowerOfTwo(const SCEV *S, bool OrZero = false,
                               bool OrNegative = false);
 
+  /// Check that memory access offsets in S are multiples of M.  Assumptions
+  /// records the runtime predicates under which S is a multiple of M.
```
```suggestion
  /// Check that memory access offsets in \p S are multiples of \p M.  \p Assumptions
  /// records the runtime predicates under which \p S is a multiple of \p M.
```

I also think that, if I understand it correctly, `S` no longer needs to be a memory access.

sebpop replied:
Done.

kasuga-fj commented on llvm/test/Analysis/DependenceAnalysis/DifferentOffsets.ll:
```diff
@@ -0,0 +1,192 @@
+; CHECK-LABEL: 'multidim_accesses'
+; CHECK-NEXT:  Src: store i32 1, ptr %idx0, align 4 --> Dst: store i32 1, ptr %idx0, align 4
+; CHECK-NEXT:    da analyze - none!
+; CHECK-NEXT:  Src: store i32 1, ptr %idx0, align 4 --> Dst: store i32 1, ptr %idx1, align 4
+; CHECK-NEXT:    da analyze - consistent output [0 0 0|<]!
```
I guess this is a case I commented on before, and I'm still not sure if this IR is valid (i.e., should we consider such a case?). However, if this is a valid input, I think this result is wrong. If I understand correctly, this loop is something like the following:

```c
for (i = 0; i < 256; i++) {
  for (j = 0; j < 256; j++) {
    for (k = 0; k < 256; k++) {
      int *idx0 = (int *)((long long *)(A) + 256*256*i + 256*j + k);
      *idx0 = 1;
      int *idx1 = (int *)((int *)(A) + 256*256*i + 256*j + k);
      *idx1 = 1;
    }
  }
}
```

For example, the memory location accessed by the first store when `(i, j, k) = (0, 0, 1)` is the same as that accessed by the second store when `(i, j, k) = (0, 0, 2)`.

(Assuming this IR is legal) This needs to be fixed, but I think it's fine to address this issue in other PR. If you intend to do so, I would prefer to leave some TODO/FIXME comments.

sebpop replied:
I added a fixme comment like this:
```
 ; CHECK-NEXT:  Src: store i32 1, ptr %idx0, align 4 --> Dst: store i32 1, ptr %idx0, align 4
 ; CHECK-NEXT:    da analyze - none!
 ; CHECK-NEXT:  Src: store i32 1, ptr %idx0, align 4 --> Dst: store i32 1, ptr %idx1, align 4
+; FIXME: the dependence distance is not constant. Distance vector should be [* * *|<]!
 ; CHECK-NEXT:    da analyze - consistent output [0 0 0|<]!
 ; CHECK-NEXT:  Src: store i32 1, ptr %idx1, align 4 --> Dst: store i32 1, ptr %idx1, align 4
 ; CHECK-NEXT:    da analyze - none!
 ;
+; for (i = 0; i < 256; i++)
+;   for (j = 0; j < 256; j++)
+;      for (k = 0; k < 256; k++) {
+;         int *idx0 = (int *)((long long *)(A) + 256*256*i + 256*j + k);
+;         *idx0 = 1;
+;         int *idx1 = (int *)((int *)(A) + 256*256*i + 256*j + k);
+;         *idx1 = 1;
+;      }
```

kasuga-fj commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -10971,6 +10971,52 @@ bool ScalarEvolution::isKnownToBeAPowerOfTwo(const SCEV *S, bool OrZero,
+bool ScalarEvolution::isKnownMultipleOf(
+    const SCEV *S, uint64_t M,
+    SmallVectorImpl<const SCEVPredicate *> &Assumptions) {
+  if (M == 0)
+    return false;
```
nit: What if `S` is also 0? This is a very special case, so I don't think we need to handle such a case carefully for now, but I would prefer to leave some TODO/FIXME comments for such a case. (For example, I think it's also reasonable to append a predicate for `S == 0` to `Assumptions` and return true here.)

Meinersbur replied:
`S==0` in the context of getMinusSCEV would mean an eltSize of 0, which means memory will never be accessed which means there never is a dependency. But there isn't necessarily this context for the caller of `isKnownMultipleOf`.

This function can always return false, since it is returns true of if it is a *known* multiple. The case `0/0` is still questionable, so it is safe to return false here.

kasuga-fj reviewed (commented):
The code looks almost good to me. By the way, what are you going to do for #116630 ? I personally think it's also fine to put it into this PR if it is more convenient for you.

Meinersbur commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -10971,6 +10971,52 @@ bool ScalarEvolution::isKnownToBeAPowerOfTwo(const SCEV *S, bool OrZero,
+  if (M == 1)
+    return true;
+
+  // Recursively check AddRec operands.
```
Consider adding a comment in the sense of:
> An AddRecExpr is a multiple of M if it starts with a multiple of M and at every steps only adds multiples of M.

Meinersbur commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -10971,6 +10971,52 @@ bool ScalarEvolution::isKnownToBeAPowerOfTwo(const SCEV *S, bool OrZero,
+
+  // Detect redundant predicates.
+  for (auto *A : Assumptions)
+    if (A->implies(P, *this))
```
If P is already implied by existing assumptions, shouldn't `isKnownPredicate(ICmpInst::ICMP_EQ, SmodM, Zero)` already have return true?

sebpop replied:
No.

`isKnownPredicate(ICmpInst::ICMP_EQ, SmodM, Zero)` returns true when `"S % M == 0" is known at compile time.`

`Assumptions` keeps constraints that cannot be disambiguated at compile time. It may be the case that we have already recorded `P` in `Assumptions` and this is a second time we see `P`.

Meinersbur commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -10971,6 +10971,52 @@ bool ScalarEvolution::isKnownToBeAPowerOfTwo(const SCEV *S, bool OrZero,
+      return true;
+
+  // Only record non-redundant predicates.
+  Assumptions.push_back(P);
```
Should there be more predicate normalizations taking place than just SCEVAddRecExpr? For instance, SCEVZeroExtendExpr is a multiple of M iff its argument is s multiple of M. It's OK if this is out-of-scope of this PR, but I would expect a TODO. Possibly these normalization should take place in getComparePredicate, just like getSCEV normalizes SCEVExprs.

sebpop replied:
I added a TODO comment just before we start building the runtime predicate:
```
+  // TODO: Also check other SCEV expressions, i.e., SCEVAddRecExpr, etc.
+
   // Basic tests have failed.
   // Check "S % M == 0" at compile time and record runtime Assumptions.
   auto *STy = dyn_cast<IntegerType>(S->getType());
```

sebpop commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3593,22 +3610,20 @@ DependenceInfo::depends(Instruction *Src, Instruction *Dst) {
-  assert(isLoadOrStore(Dst) && "instruction is not load or store");
-  Value *SrcPtr = getLoadStorePointerOperand(Src);
-  Value *DstPtr = getLoadStorePointerOperand(Dst);
+  const MemoryLocation &DstLoc = MemoryLocation::get(Dst);
```
The asserts are redundant with the conditions and early exit at line 3610.

sebpop commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3617,30 +3632,80 @@ DependenceInfo::depends(Instruction *Src, Instruction *Dst) {
+  }
+
+  uint64_t EltSize = SrcLoc.Size.toRaw();
+  assert(EltSize == DstLoc.Size.toRaw() && "Array element size differ");
```
Removed this assert as it is redundant.

sebpop committed:
update comments for last wave of reviews

kasuga-fj commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3617,30 +3632,80 @@ DependenceInfo::depends(Instruction *Src, Instruction *Dst) {
-
-  FullDependence Result(Src, Dst, PossiblyLoopIndependent, CommonLevels);
-  ++TotalArrayPairs;
+  if (DstLoc.Size != SrcLoc.Size) {
```
If I understand correctly, this check seems to be insufficient. For example, I think it should also be rejected if both of them are a `BeforeOrAfterPointer`, which is defined in a `LocationSize`.

kasuga-fj reviewed (commented):
Could you please update the PR title and description?

sebpop committed:
also reject before or after pointer

kasuga-fj commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3617,30 +3633,79 @@ DependenceInfo::depends(Instruction *Src, Instruction *Dst) {
-
-  FullDependence Result(Src, Dst, PossiblyLoopIndependent, CommonLevels);
-  ++TotalArrayPairs;
+  if (DstLoc.Size != SrcLoc.Size || DstLoc.Size.mayBeBeforePointer() ||
+      SrcLoc.Size.mayBeBeforePointer()) {
```
```suggestion
  if (DstLoc.Size != SrcLoc.Size || DstLoc.Size.isPrecise() ||
      SrcLoc.Size.isPrecise()) {
```

I don't know much about `LocationSize`, but as reading the code, using `isPrecise` seems to be suitable to me.

kasuga-fj reviewed (approved):
I still think there may be some corner cases, but this patch is taking things in the right direction. There might be some minor details that should be corrected, but for now, it looks good to me.

It's better to wait a while for responses from others who left comments.

sebpop committed:
use LocationSize.isPrecise

Co-authored-by: Ryotaro Kasuga <kasuga.ryotaro@fujitsu.com>

sebpop committed:
remove Assumptions.empty branch

sebpop commented:
Hi,
I intend to leave this patch up for review to collect more comments until Monday May 19.
Let me know about other changes before merging.
